lib/widgets/message_list.dart

```diff
@@ -250,6 +247,30 @@ class _MessageListState extends State<MessageList> with PerAccountStoreAwareStat
       // TODO: Offer `ScrollViewKeyboardDismissBehavior.interactive` (or
       //   similar) if that is ever offered:
       //     https://github.com/flutter/flutter/issues/57609#issuecomment-1355340849
+      findChildIndexCallback: (Key key) {
+        // To preserve state across rebuilds for individual [MessageItem]
+        // widgets as the size of [MessageListView.items] changes we need
+        // to match old widgets by their key to their new position in
+        // the list.
+        //
+        // The keys are of type [ValueKey] with a value of [Message.id]
+        // and here we use a O(log n) binary search method. This could
+        // be improved but for now it only triggers for materialized
+        // widgets. As a simple test, scrolling through All Messages in
+        // CZO on a Pixel 5, this only runs about 10 times per rebuild
+        // and the timing for each call is measured in microseconds.
+        //
+        // Non-message items (e.g., start and end markers) that do not
+        // have state that needs to be preserved have not been given keys
+        // and will not trigger this callback.
+        final valueKey = key as ValueKey;
+        final index = model!.findItemWithMessageId(valueKey.value);
+        if (index == -1) {
+          return null;
+        }
+        // Note `itemBuilder` pulls from `model.items` in reverse order.
+        return length - index - 1;
```

Comment:
Clearest to group these terms in the same way as in `itemBuilder`:
```suggestion
        return length - 1 - index;
```

That way when comparing the two, "length - 1" can be visually taken together as a single common quantity with a minimum of mental algebra.